File: bot.js
const url = "http://localhost:11434/api/generate";

let chatBox = null; // To hold reference to chat box
let userInput = null; // To hold reference to user input

// Define appendMessage globally
function appendMessage(role, text) {
    const messageDiv = document.createElement('div');
    messageDiv.className = `message ${role}`;
    messageDiv.textContent = text;
    chatBox.appendChild(messageDiv);
    chatBox.scrollTop = chatBox.scrollHeight;
    return messageDiv;
}

async function llama3(prompt, context = null) {
    const data = {
        model: "zlashAi",
        prompt: prompt,
        stream: true,
    };

    if (context) {
        data.context = context;
    }

    const headers = {
        'Content-Type': 'application/json'
    };

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(data)
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = ""; // Buffer to accumulate incoming text
        let newContext = context ? context : [];
        let botMessageDiv = null;

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            const decodedLine = decoder.decode(value, { stream: true });
            console.log('Received chunk:', decodedLine); // Debugging output
            buffer += decodedLine; // Accumulate buffer

            try {
                const responseData = JSON.parse(decodedLine);

                if (responseData.response) {
                    buffer += responseData.response;
                    const words = buffer.split(/\s+/); // Split buffer into words

                    if (!botMessageDiv) {
                        botMessageDiv = appendMessage('bot', '');
                    }

                    // Update the message div with completed words
                    const lastWord = words.pop(); // Get the last word
                    botMessageDiv.textContent += lastWord + ' ';
                    
                    buffer = ''; // Clear buffer after updating
                }

                if (responseData.context) {
                    newContext = responseData.context;
                }

                if (responseData.done) {
                    console.log('Response done.'); // Debugging output
                    break;
                }
            } catch (e) {
                console.error("JSON Decode Error:", e);
            }
        }

        return { content: buffer, newContext };
    } catch (e) {
        console.error("Request Error:", e);
    }
}

document.addEventListener('DOMContentLoaded', () => {
    chatBox = document.getElementById('chat-box');
    userInput = document.getElementById('user-input');

    async function sendMessage() {
        const userMessage = userInput.value.trim();
        if (!userMessage) return;

        appendMessage('user', userMessage);
        userInput.value = '';

        await llama3(userMessage);
    }

    userInput.addEventListener('keydown', (event) => {
        if (event.key === 'Enter') {
            if (!event.shiftKey) {
                event.preventDefault();
                sendMessage();
            } else {
                userInput.value += '\n';
            }
        }
    });

    window.sendMessage = sendMessage;
});

File: bot.py
import requests
import json
import sys
import time

url = "http://localhost:11434/api/generate"

def llama3(prompt, context=None):
    data = {
        "model": "zlashAi",
        "prompt": prompt,
    }
    
    if context:
        data["context"] = context
    
    headers = {
        'Content-Type': 'application/json'
    }
    
    try:
        response = requests.post(url, headers=headers, json=data, stream=True)
        
        content = ""
        new_context = context if context else [] 
        
        for line in response.iter_lines():
            if line:
                decoded_line = line.decode('utf-8')
                try:
                    response_data = json.loads(decoded_line)
                    
                    if 'response' in response_data and response_data['response']:
                        content += response_data['response']
                        
                        sys.stdout.write("\rzlashAi: " + content)
                        sys.stdout.flush()
                
                    if 'context' in response_data:
                        new_context = response_data['context']
                    
                    if response_data.get('done', False):
                        sys.stdout.write("\n")  
                        sys.stdout.flush()
                        break
                
                except json.JSONDecodeError as e:
                    print("JSON Decode Error:", e)
    
    except requests.RequestException as e:
        print("Request Error:", e)
    
    return content, new_context

def chat_loop():
    previous_context = []  # Initialize with an empty context or context from previous responses
    
    print("Welcome to the chat! Type 'exit' to end the conversation.")
    
    while True:
        user_input = input("You: ")
        
        if user_input.lower() == 'exit':
            print("Ending the conversation. Goodbye!")
            break
        
        # Print initial prompt and then get the response
        sys.stdout.write("zlashAi: ")
        sys.stdout.flush()
        
        response_content, new_context = llama3(user_input, context=previous_context)
        
        # Update context for future requests
        previous_context = new_context

# Run the chat loop
chat_loop()
